### PR TITLE
Fix generating two different BIR files for cache and --bir-dump-file flag

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/28289

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java [hunk 1]
```diff
@@ -405,15 +403,8 @@ private static void cacheBIR(ModuleContext moduleContext) {
...
-        try {
-            byte[] pkgBirBinaryContent = PackageFileWriter.writePackage(
-                    moduleContext.bLangPackage.symbol.birPackageFile);
-            birContent.writeBytes(pkgBirBinaryContent);
-            moduleContext.compilationCache.cacheBir(moduleContext.moduleName(), birContent);
-        } catch (IOException e) {
-            // This path may never be executed
-            throw new RuntimeException("Failed to convert BIR model to a byte array", e);
-        }
+        birContent.writeBytes(moduleContext.bLangPackage.symbol.birPackageFile.pkgBirBinaryContent);
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:406]: Looks like the BIR is already converted to binary before executing this line? I think the project API should have control over generating the binary representation.

**Review by sameerajayasoma — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java (hunk 1 above)
Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:406]: Yes. We do at Birgen. We keep the binary at BPackageSymbol. I think this was done mainly to ensure serialization done only once and reuse binary content in different places ( bir-dump-file flag, bir cache).

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:406]: https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java#L265

Comment (sameerajayasoma) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:406]: Can we do the serialization only when it is required? On-demand basically.

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:406]: We can do it. We can cache it the same way (so we do not need to serialize it multiple times). I'll update the PR.

**Comment by warunalakshitha** [comment]
I am closing this since my impl has changes. Will send a new PR

**Pull request closed without merging**